pkg/controller/directpvmigrationprogress/directvolumemigrationprogress_controller.go

```diff
@@ -0,0 +1,368 @@
+	}
+
+	pvProgress.Status.SetReady(!pvProgress.Status.HasCriticalCondition(), "The progress is available")
+	if pvProgress.Status.HasCriticalCondition() {
+		pvProgress.Status.PodPhase = ""
+	}
+
+	pvProgress.Status.EndStagingConditions()
+
+	pvProgress.MarkReconciled()
+	err = r.Update(context.TODO(), pvProgress)
+	if err != nil {
+		log.Trace(err)
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	if !pvProgress.Status.IsReady() {
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	// we will requeue this every 5 seconds
+	return reconcile.Result{Requeue: true, RequeueAfter: time.Second * 5}, nil
+}
+
+func (r *ReconcileDirectVolumeMigrationProgress) reportContainerStatus(pvProgress *migapi.DirectVolumeMigrationProgress, containerName string) error {
+	podRef := pvProgress.Spec.PodRef
+	ref := pvProgress.Spec.ClusterRef
+
+	// NotSet
+	if !migref.RefSet(ref) {
+		pvProgress.Status.SetCondition(migapi.Condition{
+			Type:     InvalidClusterRef,
+			Status:   migapi.True,
+			Reason:   NotSet,
+			Category: migapi.Critical,
+			Message:  "The spec.clusterRef must reference name and namespace of a valid `MigCluster",
+		})
+		return nil
+	}
+
+	cluster, err := migapi.GetCluster(r, ref)
+	if err != nil {
+		return liberr.Wrap(err)
+	}
+
+	// NotFound
+	if cluster == nil {
+		pvProgress.Status.SetCondition(migapi.Condition{
+			Type:     InvalidClusterRef,
+			Status:   migapi.True,
+			Reason:   NotFound,
+			Category: migapi.Critical,
+			Message: fmt.Sprintf("The spec.clusterRef must reference a valid `MigCluster` %s",
+				path.Join(ref.Namespace, ref.Name)),
+		})
+		return nil
+	}
+
+	// Not ready
+	if !cluster.Status.IsReady() {
+		pvProgress.Status.SetCondition(migapi.Condition{
+			Type:     ClusterNotReady,
+			Status:   migapi.True,
+			Reason:   NotReady,
+			Category: migapi.Critical,
+			Message: fmt.Sprintf("The `MigCluster` spec.ClusterRef %s is not ready",
+				path.Join(ref.Namespace, ref.Name)),
+		})
+	}
+
+	pod, err := r.Pod(cluster, podRef)
+	switch {
+	case errors.IsNotFound(err):
+		// handle not found and return
+		pvProgress.Status.SetCondition(migapi.Condition{
+			Type:     InvalidPod,
+			Status:   migapi.True,
+			Reason:   NotFound,
+			Category: migapi.Critical,
+			Message: fmt.Sprintf("The spec.podRef %s must reference a valid `Pod` ",
+				path.Join(podRef.Namespace, podRef.Name)),
+		})
+		return nil
+	case err != nil:
+		return liberr.Wrap(err)
+	}
+
+	var containerStatus *kapi.ContainerStatus
+	for _, c := range pod.Status.ContainerStatuses {
+		if c.Name == containerName {
+			containerStatus = &c
+		}
+	}
+
+	if containerStatus == nil {
+		pvProgress.Status.SetCondition(migapi.Condition{
+			Type:     InvalidPod,
+			Status:   migapi.True,
+			Reason:   NotFound,
+			Category: migapi.Critical,
+			Message: fmt.Sprintf("The spec.podRef %s must reference a `Pod` with container name %s",
+				path.Join(podRef.Namespace, podRef.Name), containerName),
+		})
+		return nil
+	}
+
+	switch {
```

Comment:
need to handle waiting status as a followup I think

Reply:
ack